### PR TITLE
Panzer: Refactor interpolation operators

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/panzer 

## Motivation
- Panzer: Move assembly of interpolation operators from MiniEM to disc-fe. Refactor to make it easier for apps to use it.
- Teko Utilities: add a matrix writer
- Panzer dof-mgr: remove Epetra dependency from test and re-enable it
- Panzer disc-fe: re-enable some tests
- Panzer project_field: fix out-of-bounds access
- Panzer MiniEM: Fix p-multigrid setup for small problems

### CODE REVIEW

**Comment by rppawlo** [comment]
@cgcgcg - thanks for moving this! Do you have a corresponding unit test that we could move into disc-fe? If testing requires the stk subpackage, then it is fine to leave in mini-em. Just hoping we could have better coverage in each subpackage so that testing is not reliant on downstream subpackages when possible.

File: packages/teko/src/Teko_Utilities.cpp [hunk 1]
```diff
@@ -2964,4 +2965,40 @@ std::string formatBlockName(const std::string &prefix, int i, int j, int nrow) {
...
+        rcp_dynamic_cast<const Epetra_CrsMatrix>(eOp->epetra_op(), true);
+    EpetraExt::BlockMapToMatrixMarketFile(("rowmap_" + s).c_str(), crsOp->RowMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("colmap_" + s).c_str(), crsOp->ColMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("domainmap_" + s).c_str(), crsOp->DomainMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("rangemap_" + s).c_str(), crsOp->RangeMap());
+    EpetraExt::RowMatrixToMatrixMarketFile(s.c_str(), *crsOp);
+  }
+#endif
+  else {
+    TEUCHOS_ASSERT(false);
```
Comment (MalachiTimothyPhillips) [inline packages/teko/src/Teko_Utilities.cpp:3000]: Might be worth adding some helpful error message here? A user could inadvertently call this with a `Teko::BlockedLinearOp`, right?

**Review by MalachiTimothyPhillips — COMMENTED** [review]
Teko related changes LGTM.

**Comment by cgcgcg** [comment]
@rppawlo Can this be merged?

File: packages/panzer/mini-em/example/BlockPrec/CMakeLists.txt [hunk 2]
```diff
@@ -292,12 +287,12 @@ IF (NOT Kokkos_ENABLE_CUDA AND ${PACKAGE_NAME}_ENABLE_ML AND MueLu_ENABLE_Epetra
...
-    NAME "Darcy_ML"
-    ARGS "--solver=ML --numTimeSteps=1 --linAlgebra=Epetra --inputFile=darcyHex.xml  --x-elements=15  --y-elements=15 --z-elements=15"
-    NUM_MPI_PROCS 4
-    )
+  # TRIBITS_ADD_TEST(
+  #   BlockPrec
+  #   NAME "Darcy_ML"
+  #   ARGS "--solver=ML --numTimeSteps=1 --linAlgebra=Epetra --inputFile=darcyHex.xml  --x-elements=15  --y-elements=15 --z-elements=15"
+  #   NUM_MPI_PROCS 4
+  #   )
```
Comment (rppawlo) [inline packages/panzer/mini-em/example/BlockPrec/CMakeLists.txt:295]: @cgcgcg - looks like the Darcy test is now disabled? Just wanted to make sure that was intentional.

**Review by rppawlo — APPROVED** [review]
@cgcgcg - looks good to me! Thanks!

File: packages/panzer/mini-em/example/BlockPrec/CMakeLists.txt (hunk 2 above)
Comment (cgcgcg) [reply packages/panzer/mini-em/example/BlockPrec/CMakeLists.txt:295]: Yes, the refactored interpolations have some issues when used with Epetra, but I don't see that as a major issue.

**Pull request merged**